### ISSUE TITLE
Level Generation weird behavior with very large submarines

### ISSUE THREAD

**Issue by Ant1V1RU5** [Regalis11/Barotrauma#2772]
After some missions on my new large submarine, i noticed some strange behavior from the Level Generation. So i collected some data for you to analyse, including crashlogs.

**Submarine used :** https://steamcommunity.com/sharedfiles/filedetails/?id=2083930490&searchtext= 
**Size :** 189x52m
**Difficulty :** Hellish (100%)
**Mode :** Sandbox
**Multiplayer :** Yes

**Seeds -- Notes** (based on 30 samples)

**tuToQGQB** -- Impossible to dive further / Wall too narrow
**gvp6C4Dj** -- Impossible to dive further / Wall too narrow
**i6g6J3Jd** -- Wall almost too narrow, but with dexterity the sub can pass trought
**mMT74fgE** -- Impossible to dive further / Wall too narrow / Alien ruin directly at the right of the submarine
**1TGmL0nw** -- Wall almost too narrow, but with dexterity the sub can pass trought

**s7lrLp59** -- Almost no walls / Minerals very concentrated at one point / Alien Ruin directly at the left of the ship touching it
**7CjxD0hF** -- Almost no walls / Minerals very concentrated at one point / Alien Ruin directly at the left of the ship touching it
**vxc8tVuj** -- Almost no walls / Minerals very concentrated at one point
**Ef556KTm** -- Almost no walls / Minerals very concentrated at one point / Outpost extremely close to the starting area
**VxtBhuXH** -- Almost no walls / Minerals very concentrated at one point / Outpost extremely close to the starting area
**HYjD2SUO** -- Almost no walls / Minerals very concentrated at one point / Outpost extremely close to the starting area

**aEOoZgjP** -- Crash Serveur
**jI3unDao** -- Crash Serveur
**VZpKIoqC** -- Crash Serveur
**nmgPiQWv** -- Crash Server
**ocIWdVlP** -- Crash Server

**IY2kjxu4** -- Ship stuck in walls at the start
**2LwPNHh2** -- Ship stuck in walls at the start

**hGcO3hZe** -- Ship stuck in outpost at the start
**2oJ8kZd0** -- Ship stuck in outpost at the start

**t2tx021W** -- Level Perfectly Fine
**1JSGgkYe** -- Level Perfectly Fine
**puuJRXaC** -- Level Perfectly Fine
**UwupEAPA** -- Level Perfectly Fine
**aMMHvpuB** -- Level Perfectly Fine
**4fBE79Lp** -- Level Perfectly Fine
**dIG7dkLx** -- Level Perfectly Fine
**WJkJYTSl** -- Level Perfectly Fine
**wGcYfCdW** -- Level Perfectly Fine
**BL7mbTtz** -- Level Perfectly Fine

**Crashlogs and Pictures :** 
[aEOoZgjP.log](https://github.com/Regalis11/Barotrauma/files/4581736/aEOoZgjP.log)
[jI3unDao.log](https://github.com/Regalis11/Barotrauma/files/4581737/jI3unDao.log)
[nmgPiQWv.log](https://github.com/Regalis11/Barotrauma/files/4581738/nmgPiQWv.log)
[ocIWdVlP.log](https://github.com/Regalis11/Barotrauma/files/4581733/ocIWdVlP.log)
[VZpKIoqC.log](https://github.com/Regalis11/Barotrauma/files/4581734/VZpKIoqC.log)

![2LwPNHh2](https://user-images.githubusercontent.com/62913364/81083429-ded6b980-8ef4-11ea-8950-831c67fce97b.png)
![2oJ8kZd0](https://user-images.githubusercontent.com/62913364/81083435-e007e680-8ef4-11ea-9233-cbb73f5af88e.png)
![hGcO3hZe](https://user-images.githubusercontent.com/62913364/81083438-e1d1aa00-8ef4-11ea-8a8d-35c2d23e47d0.png)
![IY2kjxu4](https://user-images.githubusercontent.com/62913364/81083442-e26a4080-8ef4-11ea-81cb-9f55029ddfaf.png)

**Ship Hull and Wall in editor :** 

![Hull + Walls in editor](https://user-images.githubusercontent.com/62913364/81083560-06c61d00-8ef5-11ea-8b6f-a9d6d5ae1dae.png)

**Edit :** Adjusted crashlogs position on the post

**Comment by Regalis11** [comment]
Wow, that's a really impressive sub, both visually and technically!

All of these issues seem to have been caused by the level generator leaving some "margin" at the sides of the level, where the outpost or the main path won't be generated. With a sub this large margin would sometimes get so wide, that the start and end of the path would be placed right next to each other, or in some cases there would be no space for the main path at all, which would cause a crash.

I added some measures in https://github.com/Regalis11/Barotrauma-development/commit/89dd695926c24539699ad9fce129bcc3847f5a10 to prevent this from happening. Even though after those changes the game shouldn't crash and those weird levels with practically no walls should not generate, I'm afraid the level generation might not always be able to guarantee a sub this big will be able to get through the level. I think we'll need to add some kind of a soft cap for the submarine size and display a warning when it goes over that limit.

**Comment by Ant1V1RU5** [comment]
The soft cap seems to be a really good idea to prevent submakers to do overlarge ship without knowing it (well i didn't know how large mine will be). However, allowing us to play on massive vessels of destruction (to a certain degree of course) is something i hope get smoothly integrated in the game.

I am glad these data were useful to you :)

**Comment by Rokvach** [comment]
All the seeds aren't useful anymore as the level generation was changed, but I went through quite a bit of different levels with the Rogue Trader sub. 

I did notice that there are still some levels which are just slightly too narrow for the sub to pass, but there are also a lot of levels where the sub doesn't have any problems navigating through. 

Should we widen the max path width a bit more or just leave it? @Regalis11
